lib/diaspora_federation/entities/account_migration.rb

```diff
@@ -23,16 +25,25 @@ class AccountMigration < Entity
       #   @return [String] signature
       property :signature, :string, default: nil
 
-      # @return [String] string representation of this object
-      def to_s
-        "AccountMigration:#{author}:#{profile.author}"
+      # @!attribute [r] old_user
+      #   Optional attribute which is used to indicate old diaspora* ID when author attribute contains new diaspora* ID
+      #   @return [String] old_user
+      property :old_user, :string, optional: true
+
+      # Returns diaspora* ID of the old person depending on the format of the message
+      # @return [String] diaspora* ID of the old person
+      def old_user_id
```

Comment:
Currently our property accessors in the entity classes have two responsibilities:
- they act as public API methods for the application to get entity properties
- they act as data source for serialization routines

In some cases, like this one, it may be useful to have a possibility to implement different behaviors for these two usages.

Currently `enriched_properties` provides the ability to change the serialized data in arbitrary way. And therefore we can change the way properties are serialized. However there are some problems with this approach. First it is not obvious that "enrich" is something that replaces the parameter. It looks hackish that we "enrich" some value by nullifying it. Also this way we have to store the unchanged value and only afterwards apply some changes instead of directly reading a correct value for serialization in the first place. And this feels like a more complex behavior, it's harder to follow.

What we can do instead is create a separate method for the serialization responsibility and make a call to respective public API method a default behavior. It can be easily achieved like this:
```ruby
diff --git a/lib/diaspora_federation/entity.rb b/lib/diaspora_federation/entity.rb
index 30549b3..d8c010f 100644
--- a/lib/diaspora_federation/entity.rb
+++ b/lib/diaspora_federation/entity.rb
@@ -264,10 +264,14 @@ module DiasporaFederation
     # @return [Hash] hash with all properties
     def properties
       self.class.class_props.keys.each_with_object({}) do |prop, hash|
-        hash[prop] = public_send(prop)
+        hash[prop] = fetch_property_for_serialization(prop)
       end
     end
 
+    def fetch_property_for_serialization(prop)
+      public_send(prop)
+    end
+
     def normalized_properties
       properties.map {|name, value| [name, normalize_property(name, value)] }.to_h
     end

```

When it is done we can override the default behavior when we want so. Here is an example for the AM entity:
```ruby
      def fetch_property_for_serialization(prop)
        case prop
        when :old_identity
          @old_identity  
        else
          super
        end
      end
```

We can also use the same approach for signatures serialization instead of the current way we are doing it right now (using enriched properties). `enriched_properties` are still useful though, e.g. for adding the `addtional_data` to the serialization hash.

This would allow us to have an adequate public API method and also a straightforward serialization behavior. And therefore we can store the `old_identity` value only when it is mandatory.

Also there are some other possible ways how we can implement the override of the serialization getter. E.g. we can make some `serialization_accessor:` parameter for properties DSL or checking for some specific names for serialization getter, like searching for `#{prop}_for_serialization` method to fetch parrameter value and fall back to `public_send(prop)` in case it is missing. In this comment I just showed the simpliest possible implementation.

At the same time, I think that allowing `old_identity` to be there always even when it is not required is okay. And I agree that relying on `author` to decide which keys to sign/verify signature with is a better idea.

What do you think about this?

Reply:
Yes, you're right. Setting something to `nil` isn't really "enrich" ;) But I think for everything else we currently use `enriched_properties` (signatures for example), it's OK.

While adding a method for serialization could have some advantaged, I think for now it would only add another layer between the properties and serialization. Another place where something could be changed. So I'm not sure if it's worth the additional complexity, when we don't need it.

If we need it, I would probably prefer a special method name to get a property for serialization, because it looks the easiest to me.

But if we agree that we allow `old_identity` always, we don't need to change `enriched_properties` now, so this would be out of scope for this PR. So we can do this refactoring in another PR, or just wait until we really need this additional layer.

Reply:
I see that you like the solution with less possible changes and I'm okay with that :) We can always add this feature later if we have stronger reasons for this.